### DISCUSSION TITLE
Is MsgPropose/MsgReadIndex silently ignored when the node role is candidate?

### DISCUSSION THREAD

**Discussion by tisonkun** [tikv/raft-rs#524] [Q&A]
Application read and write generally go into `MsgPropose` or `MsgReadIndex`. While I don't have ideas on how to sort out inbound messages, I notice that MsgPropose/MsgReadIndex will be silently ignored when the node role is in candidate.

Even I can store the node role and update on `Ready`, I `step` messages in batch and handle Ready at the end. So it's possible that node role switch to candidate intermediately.

Shall I force handle Ready and check node role before `step` MsgPropose/MsgReadIndex? Or is there other best practice to properly handle application issues read and write (MsgPropose/MsgReadIndex)?

cc @BusyJay

**Comment by BusyJay** [comment]
TiKV will check pending entries and pending read index count to detect if a request is accepted. Though it's more appropriate to return an error instead.
